plugins/catalog-backend-module-geoserver-data/src/GeoserverDataProvider.ts

```diff
@@ -0,0 +1,314 @@
+import {
+  EntityProvider,
+  EntityProviderConnection,
+} from '@backstage/plugin-catalog-node';
+import {
+  LoggerService,
+  SchedulerServiceTaskRunner,
+} from '@backstage/backend-plugin-api';
+// @ts-ignore
+import { GeoServerRestClient } from 'geoserver-rest-client';
+import {
+  ANNOTATION_LOCATION,
+  ANNOTATION_ORIGIN_LOCATION,
+  Entity,
+  stringifyEntityRef,
+} from '@backstage/catalog-model';
+
+export class GeoserverDataProvider implements EntityProvider {
+  private connection?: EntityProviderConnection;
+
+  constructor(
+    private uri: string,
+    private username: string,
+    private password: string,
+    private taskRunner: SchedulerServiceTaskRunner,
+    private logger: LoggerService,
+  ) {}
+
+  getProviderName(): string {
+    return `geoserver-data-${this.uri}`;
+  }
+
+  async connect(connection: EntityProviderConnection) {
+    this.connection = connection;
+
+    await this.taskRunner.run({
+      id: this.getProviderName(),
+      fn: async () => {
+        await this.run();
+      },
+    });
+    await this.run();
+  }
+
+  async run() {
+    if (!this.connection) {
+      throw new Error('Not initialized');
+    }
+
+    const grc = new GeoServerRestClient(this.uri, this.username, this.password);
+
+    const workspacesObject = await grc.workspaces.getAll();
+    const workspaces = workspacesObject?.workspaces?.workspace;
+
+    if (workspaces === undefined || workspaces.length === 0) {
+      this.logger.info(
+        'No workspaces found in GeoServer. No Geoserver entities will be created.',
+      );
+      return;
+    }
+
+    const entities: Entity[] = [];
+
+    for (const workspace of workspaces || []) {
+      this.logger.debug(
+        `Found GeoServer workspace: ${workspace.name} at ${workspace.href}`,
+      );
+
+      // Add data stores
+      const dataStoresForWorkspaceObject = await grc.datastores.getDataStores(
+        workspace.name,
+      );
+      const dataStoresForWorkspace =
+        dataStoresForWorkspaceObject?.dataStores?.dataStore;
+      for (const dataStore of dataStoresForWorkspace || []) {
+        this.logger.debug(
+          `Found data store: ${dataStore.name} at ${dataStore.href} for workspace ${workspace.name}`,
+        );
+
+        const fullDataStoreObject = await grc.datastores.getDataStore(
+          workspace.name,
+          dataStore.name,
+        );
+
+        const dataStoreEntity: Entity = {
+          apiVersion: 'geoportia.se/v1alpha1',
+          kind: 'GeoserverStore',
+          metadata: {
+            name: dataStore.name,
+            namespace: workspace.name,
+            description: fullDataStoreObject?.dataStore?.description,
+            annotations: {
+              [ANNOTATION_LOCATION]: `url:${this.uri}`,
+              [ANNOTATION_ORIGIN_LOCATION]: `url:${this.uri}`,
+            },
+          },
+          spec: fullDataStoreObject?.dataStore || {},
+        };
```

Comment:
There is a lot of repetition in this file, especially surrounding this. Recommend creating a method:

`private createStoreEntity(name: string, workspace: string, description: string, spec: ?)`

and use it to create each of the entities